### PR TITLE
feat: save def locals and pc

### PR DESCRIPTION
**Pull request by JonathanStarup**
Related to #6698

* Merge `GenFunctionClass` and `GenClosureClasses` since they are 95% identical
* save all local variables of functions/closures as fields which are restored upon entry
* have a `pc` field which is currently unused

Unexpected problem: if the `copy` function is inlined, methods become too big to be valid. the `copy` function is the one responsible for creating a new frame, i.e. duplicating the current def/closure. These things have tens of local variables and its done on each non-tail call.

An overview of the function/closure class is in the documentation

### CODE REVIEW

**Comment by magnus-madsen** [comment]
I see some build errors.

> Unexpected problem: if the copy function is inlined then methods become too big to be valid.

Is this solved? What "copy" function is this?

**Comment by JonathanStarup** [comment]
> I see some build errors.
> 
> > Unexpected problem: if the copy function is inlined then methods become too big to be valid.
> 
> Is this solved? What "copy" function is this?

Ah, never rename manually. It is resolved. I've expanded the explanation a bit in the main text.

**Comment by JonathanStarup** [comment]
@magnus-madsen After running some metrics, the reason is that some functions have up to 90 local variables. Just estimating based on random programs I think the median number is maybe 10

**Comment by magnus-madsen** [comment]
> @magnus-madsen After running some metrics, the reason for potentially large methods, is that some functions have up to 90 local variables. Just estimating based on random programs The median use 0 local variables but if you don't count those, then the median is probably around 10

What functions?

Will this cause breakages? (i.e. I don't want to merge something if it breaks code).

Does disabling the optimizer change it?

**Comment by magnus-madsen** [comment]
!!!jar

**Comment by magnus-madsen** [comment]
I see functions like:

```
Vector.join$5365225
RedBlackTree.rotate$5339518
Concurrent.Channel.put$5339935
Concurrent.Channel.getHelper$5341144
```

with more than 50 local vars-- which admittedly is on the high end. Nevertheless, the JVM supports up to 65535.

So I assume the problem is with the code this PR is generating?

**Comment by magnus-madsen** [comment]
Ah-- but I guess you fix is to use a helper function -- copy -- to avoid the code blowup and then everything should be fine?

**Comment by JonathanStarup** [comment]
> Ah-- but I guess you fix is to use a helper function -- copy -- to avoid the code blowup and then everything should be fine?

yes exactly, otherwise the methods become too large

**Comment by magnus-madsen** [comment]
Do you believe this to be stable enough to merge in?

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenFunAndClosureClasses.scala [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        macc + (functionname -> JvmClass(functionname, code))
+
+      case (macc, _) =>
+        macc
+    }, _ ++ _)
+  }
+
+  private def isClosure(defn: Def): Boolean = defn.cparams.nonEmpty
+
+  private def isFunction(defn: Def): Boolean = defn.cparams.isEmpty
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenFunAndClosureClasses.scala:59]: Something about excluded middle, could define isClosure in terms of isFunction

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenFunAndClosureClasses.scala [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+
+  private sealed trait FunctionKind
+
+  private case object Function extends FunctionKind
+
+  private case object Closure extends FunctionKind
+
+  /**
+    * `(a|b)` is used to represent the function (left) or closure version (right)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenFunAndClosureClasses.scala:68]: Thanks for example 👍

**Review by magnus-madsen — APPROVED** [review]

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenFunAndClosureClasses.scala (hunk 1 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/jvm/GenFunAndClosureClasses.scala:59]: Yeah but i like it for readability. I can delete one of them if you want

**Comment by JonathanStarup** [comment]
> Do you believe this to be stable enough to merge in?

@magnus-madsen It should be stable, I'm mostly questioning performance

**Comment by magnus-madsen** [comment]
Merged! Will you update the ticket :)

**Pull request merged**